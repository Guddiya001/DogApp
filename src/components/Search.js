import React from "react";
import { useState } from "react";
import { useAppDispatch } from '../hooks'
import { fetchDogs, searchFetchDogs } from "../redux/actions";
const Search = () => {
  const [search, setSearch] = useState("");

  const searchAnimal = (event) => {
    setSearch(event);
    let timeoutId;
    if (timeoutId) {
      clearTimeout(timeoutId);
    }
    timeoutId = setTimeout(() => {

      search.length !== 0
        ? dispatch(searchFetchDogs(search))
        : dispatch(fetchDogs());
    }, 1000)

  }

  const dispatch = useAppDispatch();

  return (
    <>
      <div className="form py-3">
        <div className="form-container w-50 mx-auto">
          <input
            type="search"
            className="form-control mx-auto"
            placeholder="search"
            id="search"
            onChange={(event) => {
              searchAnimal(event.target.value)
            }}
            value={search}
            data-testid="searchAnimal"
          />
          <i className="fas fa-search"></i>
        </div>
      </div>
    </>
  );
};

export default Search;
