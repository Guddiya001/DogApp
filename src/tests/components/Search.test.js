import React from 'react';
import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Search from "../../components/Search";

const mockStore = configureStore([thunk]);

describe('Search box testing for debouncing functionality', () => {
    let store;
    let initialState

    beforeEach(() => {
        jest.clearAllMocks()
        jest.useFakeTimers()
        jest.spyOn(global, 'setTimeout')

        initialState = {
            dog: {
                "weight": {
                    "imperial": "40 - 65",
                    "metric": "18 - 29"
                },
                "height": {
                    "imperial": "22 - 26",
                    "metric": "56 - 66"
                },
                "id": 38,
                "name": "Belgian Tervuren",
                "bred_for": "Guarding, Drafting, Police work.",
                "breed_group": "Herding",
                "life_span": "10 - 12 years",
                "temperament": "Energetic, Alert, Loyal, Intelligent, Attentive, Protective",
                "reference_image_id": "B1KdxlcNX",
                "image": {
                    "id": "B1KdxlcNX",
                    "width": 645,
                    "height": 380,
                    "url": "https://cdn2.thedogapi.com/images/B1KdxlcNX.jpg"
                }
            },
            dogs: [{
                "weight": {
                    "imperial": "40 - 65",
                    "metric": "18 - 29"
                },
                "height": {
                    "imperial": "22 - 26",
                    "metric": "56 - 66"
                },
                "id": 38,
                "name": "Belgian Tervuren",
                "bred_for": "Guarding, Drafting, Police work.",
                "breed_group": "Herding",
                "life_span": "10 - 12 years",
                "temperament": "Energetic, Alert, Loyal, Intelligent, Attentive, Protective",
                "reference_image_id": "B1KdxlcNX",
                "image": {
                    "id": "B1KdxlcNX",
                    "width": 645,
                    "height": 380,
                    "url": "https://cdn2.thedogapi.com/images/B1KdxlcNX.jpg"
                }
            },
            {
                "weight": {
                    "imperial": "40 - 65",
                    "metric": "18 - 29"
                },
                "height": {
                    "imperial": "22 - 26",
                    "metric": "56 - 66"
                },
                "id": 3,
                "name": "Belgian Tervuren",
                "bred_for": "Guarding, Drafting, Police work.",
                "breed_group": "Herding",
                "life_span": "10 - 12 years",
                "temperament": "Energetic, Alert, Loyal, Intelligent, Attentive, Protective",
                "reference_image_id": "B1KdxlcNX",
                "image": {
                    "id": "B1KdxlcNX",
                    "width": 645,
                    "height": 380,
                    "url": "https://cdn2.thedogapi.com/images/B1KdxlcNX.jpg"
                }
            },
            ],
            loading: false,
        }

        store = mockStore(initialState);
    });

    it('should display snapshot', async () => {
        const { asFragment } = render(
            <Provider store={store}>
                <Search />
            </Provider>

        );
        expect(asFragment()).toMatchSnapshot()

    });

    it('should debounce a function', () => {

        const { getByTestId } = render(<Provider store={store}>
            <Search />
        </Provider>);
        const input = getByTestId("searchAnimal");

        fireEvent.change(input, { target: { value: "Belgian Tervuren" } });
        expect(setTimeout).toHaveBeenCalledTimes(1);

        fireEvent.change(input, { target: { value: "Belgian T" } });
        expect(setTimeout).toHaveBeenCalledTimes(2);

        jest.advanceTimersByTime(1000);

        waitFor(() => {
            expect(console.log).toHaveBeenCalledWith("searching for: Belgian Tervu");
        });
    });
});
