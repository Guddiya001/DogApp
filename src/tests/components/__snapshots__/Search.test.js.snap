// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyComponent should display data fetched from API 1`] = `
<DocumentFragment>
  <div
    class="form py-3"
  >
    <div
      class="form-container w-50 mx-auto"
    >
      <input
        class="form-control mx-auto"
        data-testid="searchAnimal"
        id="search"
        placeholder="search"
        type="search"
        value=""
      />
      <i
        class="fas fa-search"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Search box testing for debouncing functionality should display snapshot 1`] = `
<DocumentFragment>
  <div
    class="form py-3"
  >
    <div
      class="form-container w-50 mx-auto"
    >
      <input
        class="form-control mx-auto"
        data-testid="searchAnimal"
        id="search"
        placeholder="search"
        type="search"
        value=""
      />
      <i
        class="fas fa-search"
      />
    </div>
  </div>
</DocumentFragment>
`;
